File: web/src/data/relicRegions.ts
/**
 * Region-to-coordinate lookup for war relic discovery locations.
 * Maps the `whereFound` strings from WarRelicSpecimens to approximate
 * lat/lng coordinates for map display. Regions are intentionally
 * coarse — these are areas, not exact sites.
 */

export interface RegionCoord {
  name: string;
  lat: number;
  lng: number;
  /** Short label shown in the map popup */
  label: string;
}

/**
 * Every distinct `whereFound` value from the 64 war relic specimens,
 * mapped to a representative coordinate. When a relic has multiple
 * regions, all of them get a marker on the map.
 */
const REGION_COORDS: Record<string, RegionCoord> = {
  // ── Broad theater / region terms ──────────────────────────────
  "Eastern Theater": { name: "Eastern Theater", lat: 38.9, lng: -77.35, label: "Eastern Theater (Virginia / Maryland)" },
  "Western Theater": { name: "Western Theater", lat: 35.15, lng: -88.65, label: "Western Theater (Tennessee / Mississippi)" },
  "Eastern battlefields": { name: "Eastern battlefields", lat: 38.5, lng: -77.5, label: "Eastern Battlefields (Virginia)" },
  "Southern battlefields": { name: "Southern battlefields", lat: 33.0, lng: -87.0, label: "Southern Battlefields" },
  "Deep South sites": { name: "Deep South sites", lat: 31.0, lng: -88.0, label: "Deep South Sites" },
  "Early war sites": { name: "Early war sites", lat: 38.5, lng: -77.5, label: "Early War Sites (Virginia)" },
  "Late-war sites": { name: "Late-war sites", lat: 35.0, lng: -81.0, label: "Late-War Sites (Carolinas)" },
  "Both armies": { name: "Both armies", lat: 38.9, lng: -77.35, label: "Both Armies (Eastern Theater)" },

  // ── State-specific ────────────────────────────────────────────
  "Virginia": { name: "Virginia", lat: 38.0, lng: -79.5, label: "Virginia" },
  "Tennessee": { name: "Tennessee", lat: 35.5, lng: -86.0, label: "Tennessee" },

  // ── Camp / position types (default to Eastern Theater if no region) ──
  "Union camps": { name: "Union camps", lat: 38.9, lng: -77.35, label: "Union Camps" },
  "Union cavalry camps": { name: "Union cavalry camps", lat: 38.9, lng: -77.35, label: "Union Cavalry Camps" },
  "Union artillery camps": { name: "Union artillery camps", lat: 38.9, lng: -77.35, label: "Union Artillery Camps" },
  "Union artillery positions": { name: "Union artillery positions", lat: 38.9, lng: -77.35, label: "Union Artillery Positions" },
  "Union positions": { name: "Union positions", lat: 38.9, lng: -77.35, label: "Union Positions" },
  "New York regiment camps": { name: "New York regiment camps", lat: 43.0, lng: -75.0, label: "New York Regiment Camps" },
  "Confederate camps": { name: "Confederate camps", lat: 37.5, lng: -78.0, label: "Confederate Camps (Virginia)" },
  "Confederate cavalry camps": { name: "Confederate cavalry camps", lat: 37.5, lng: -78.0, label: "Confederate Cavalry Camps" },
  "Confederate artillery camps": { name: "Confederate artillery camps", lat: 37.5, lng: -78.0, label: "Confederate Artillery Camps" },
  "Confederate artillery positions": { name: "Confederate artillery positions", lat: 37.5, lng: -78.0, label: "Confederate Artillery Positions" },
  "Confederate positions": { name: "Confederate positions", lat: 37.5, lng: -78.0, label: "Confederate Positions" },
  "Georgia troop camps": { name: "Georgia troop camps", lat: 32.5, lng: -83.0, label: "Georgia Troop Camps" },
  "British camps": { name: "British camps", lat: 40.7, lng: -74.0, label: "British Camps (New York / Boston)" },
  "British & American camps": { name: "British & American camps", lat: 40.7, lng: -74.0, label: "British & American Camps" },
  "American camps": { name: "American camps", lat: 40.7, lng: -74.0, label: "American Camps" },
  "Continental Army camps": { name: "Continental Army camps", lat: 40.0, lng: -75.0, label: "Continental Army Camps" },
  "Officer camps": { name: "Officer camps", lat: 38.9, lng: -77.35, label: "Officer Camps" },
  "NCO camps": { name: "NCO camps", lat: 38.9, lng: -77.35, label: "NCO Camps" },
  "Camps": { name: "Camps", lat: 38.9, lng: -77.35, label: "Camp Sites" },
  "Cavalry camps": { name: "Cavalry camps", lat: 38.9, lng: -77.35, label: "Cavalry Camps" },
  "Cavalry sites": { name: "Cavalry sites", lat: 38.9, lng: -77.35, label: "Cavalry Sites" },
  "Winter quarters": { name: "Winter quarters", lat: 38.7, lng: -77.15, label: "Winter Quarters" },
  "March routes": { name: "March routes", lat: 38.9, lng: -77.35, label: "March Routes" },
  "Rifle pits": { name: "Rifle pits", lat: 38.9, lng: -77.35, label: "Rifle Pits" },
  "Riflemen camps": { name: "Riflemen camps", lat: 38.9, lng: -77.35, label: "Riflemen Camps" },
  "Trench lines": { name: "Trench lines", lat: 37.5, lng: -77.4, label: "Trench Lines (Petersburg)" },
  "Firing lines": { name: "Firing lines", lat: 38.9, lng: -77.35, label: "Firing Lines" },
  "Defensive lines": { name: "Defensive lines", lat: 38.9, lng: -77.35, label: "Defensive Lines" },
  "Fortifications": { name: "Fortifications", lat: 38.9, lng: -77.35, label: "Fortifications" },
  "Siege positions": { name: "Siege positions", lat: 37.23, lng: -77.40, label: "Siege Positions (Petersburg)" },
  "Siege sites": { name: "Siege sites", lat: 37.23, lng: -77.40, label: "Siege Sites (Petersburg)" },
  "Artillery positions": { name: "Artillery positions", lat: 38.9, lng: -77.35, label: "Artillery Positions" },
  "Battery positions": { name: "Battery positions", lat: 38.9, lng: -77.35, label: "Battery Positions" },
  "Battery sites": { name: "Battery sites", lat: 38.9, lng: -77.35, label: "Battery Sites" },
  "Battlefields": { name: "Battlefields", lat: 38.9, lng: -77.35, label: "Battlefields" },
  "Naval sites": { name: "Naval sites", lat: 36.9, lng: -76.0, label: "Naval Sites (Hampton Roads)" },
  "Frontier sites": { name: "Frontier sites", lat: 40.0, lng: -80.0, label: "Frontier Sites" },
  "Ambush sites": { name: "Ambush sites", lat: 38.9, lng: -77.35, label: "Ambush Sites" },
  "Close-quarters engagements": { name: "Close-quarters engagements", lat: 38.9, lng: -77.35, label: "Close-Quarters Engagements" },
  "Southern fortifications": { name: "Southern fortifications", lat: 33.0, lng: -87.0, label: "Southern Fortifications" },

  // ── Source / origin terms ─────────────────────────────────────
  "Imported from Britain": { name: "Imported from Britain", lat: 53.0, lng: -1.5, label: "Imported from Britain" },
};

/**
 * Default fallback coordinate when a `whereFound` string is not in the lookup.
 * Centered on the Eastern Theater (most common CW relic region).
 */
const DEFAULT_COORD: RegionCoord = {
  name: "Unknown",
  lat: 38.9,
  lng: -77.35,
  label: "Discovery Area",
};

/**
 * Given a list of `whereFound` strings from an artifact, return
 * the matching coordinates for each one. Strings that don't match
 * the lookup are skipped (no default fallback — we only show
 * regions we can confidently place on the map).
 */
export function getRegionCoords(whereFound: string[]): RegionCoord[] {
  const results: RegionCoord[] = [];
  const seen = new Set<string>();
  for (const region of whereFound) {
    const coord = REGION_COORDS[region];
    if (coord && !seen.has(coord.label)) {
      results.push(coord);
      seen.add(coord.label);
    }
  }
  return results;
}

File: android-rockhound/app/src/main/java/com/rork/rockscout/data/RelicRegions.kt
package com.rork.rockscout.data

/**
 * Region-to-coordinate lookup for war relic discovery locations.
 * Maps the `whereFound` strings from [WarRelicSpecimens] to approximate
 * lat/lng coordinates for map display. Regions are intentionally coarse —
 * these are areas, not exact sites.
 */
object RelicRegions {

    data class RegionCoord(
        val name: String,
        val lat: Double,
        val lng: Double,
        val label: String,
    )

    private val REGION_COORDS: Map<String, RegionCoord> = mapOf(
        // ── Broad theater / region terms ──────────────────────────
        "Eastern Theater" to RegionCoord("Eastern Theater", 38.9, -77.35, "Eastern Theater (Virginia / Maryland)"),
        "Western Theater" to RegionCoord("Western Theater", 35.15, -88.65, "Western Theater (Tennessee / Mississippi)"),
        "Eastern battlefields" to RegionCoord("Eastern battlefields", 38.5, -77.5, "Eastern Battlefields (Virginia)"),
        "Southern battlefields" to RegionCoord("Southern battlefields", 33.0, -87.0, "Southern Battlefields"),
        "Deep South sites" to RegionCoord("Deep South sites", 31.0, -88.0, "Deep South Sites"),
        "Early war sites" to RegionCoord("Early war sites", 38.5, -77.5, "Early War Sites (Virginia)"),
        "Late-war sites" to RegionCoord("Late-war sites", 35.0, -81.0, "Late-War Sites (Carolinas)"),
        "Both armies" to RegionCoord("Both armies", 38.9, -77.35, "Both Armies (Eastern Theater)"),

        // ── State-specific ────────────────────────────────────────
        "Virginia" to RegionCoord("Virginia", 38.0, -79.5, "Virginia"),
        "Tennessee" to RegionCoord("Tennessee", 35.5, -86.0, "Tennessee"),

        // ── Camp / position types ─────────────────────────────────
        "Union camps" to RegionCoord("Union camps", 38.9, -77.35, "Union Camps"),
        "Union cavalry camps" to RegionCoord("Union cavalry camps", 38.9, -77.35, "Union Cavalry Camps"),
        "Union artillery camps" to RegionCoord("Union artillery camps", 38.9, -77.35, "Union Artillery Camps"),
        "Union artillery positions" to RegionCoord("Union artillery positions", 38.9, -77.35, "Union Artillery Positions"),
        "Union positions" to RegionCoord("Union positions", 38.9, -77.35, "Union Positions"),
        "New York regiment camps" to RegionCoord("New York regiment camps", 43.0, -75.0, "New York Regiment Camps"),
        "Confederate camps" to RegionCoord("Confederate camps", 37.5, -78.0, "Confederate Camps (Virginia)"),
        "Confederate cavalry camps" to RegionCoord("Confederate cavalry camps", 37.5, -78.0, "Confederate Cavalry Camps"),
        "Confederate artillery camps" to RegionCoord("Confederate artillery camps", 37.5, -78.0, "Confederate Artillery Camps"),
        "Confederate artillery positions" to RegionCoord("Confederate artillery positions", 37.5, -78.0, "Confederate Artillery Positions"),
        "Confederate positions" to RegionCoord("Confederate positions", 37.5, -78.0, "Confederate Positions"),
        "Georgia troop camps" to RegionCoord("Georgia troop camps", 32.5, -83.0, "Georgia Troop Camps"),
        "British camps" to RegionCoord("British camps", 40.7, -74.0, "British Camps (New York / Boston)"),
        "British & American camps" to RegionCoord("British & American camps", 40.7, -74.0, "British & American Camps"),
        "American camps" to RegionCoord("American camps", 40.7, -74.0, "American Camps"),
        "Continental Army camps" to RegionCoord("Continental Army camps", 40.0, -75.0, "Continental Army Camps"),
        "Officer camps" to RegionCoord("Officer camps", 38.9, -77.35, "Officer Camps"),
        "NCO camps" to RegionCoord("NCO camps", 38.9, -77.35, "NCO Camps"),
        "Camps" to RegionCoord("Camps", 38.9, -77.35, "Camp Sites"),
        "Cavalry camps" to RegionCoord("Cavalry camps", 38.9, -77.35, "Cavalry Camps"),
        "Cavalry sites" to RegionCoord("Cavalry sites", 38.9, -77.35, "Cavalry Sites"),
        "Winter quarters" to RegionCoord("Winter quarters", 38.7, -77.15, "Winter Quarters"),
        "March routes" to RegionCoord("March routes", 38.9, -77.35, "March Routes"),
        "Rifle pits" to RegionCoord("Rifle pits", 38.9, -77.35, "Rifle Pits"),
        "Riflemen camps" to RegionCoord("Riflemen camps", 38.9, -77.35, "Riflemen Camps"),
        "Trench lines" to RegionCoord("Trench lines", 37.5, -77.4, "Trench Lines (Petersburg)"),
        "Firing lines" to RegionCoord("Firing lines", 38.9, -77.35, "Firing Lines"),
        "Defensive lines" to RegionCoord("Defensive lines", 38.9, -77.35, "Defensive Lines"),
        "Fortifications" to RegionCoord("Fortifications", 38.9, -77.35, "Fortifications"),
        "Siege positions" to RegionCoord("Siege positions", 37.23, -77.40, "Siege Positions (Petersburg)"),
        "Siege sites" to RegionCoord("Siege sites", 37.23, -77.40, "Siege Sites (Petersburg)"),
        "Artillery positions" to RegionCoord("Artillery positions", 38.9, -77.35, "Artillery Positions"),
        "Battery positions" to RegionCoord("Battery positions", 38.9, -77.35, "Battery Positions"),
        "Battery sites" to RegionCoord("Battery sites", 38.9, -77.35, "Battery Sites"),
        "Battlefields" to RegionCoord("Battlefields", 38.9, -77.35, "Battlefields"),
        "Naval sites" to RegionCoord("Naval sites", 36.9, -76.0, "Naval Sites (Hampton Roads)"),
        "Frontier sites" to RegionCoord("Frontier sites", 40.0, -80.0, "Frontier Sites"),
        "Ambush sites" to RegionCoord("Ambush sites", 38.9, -77.35, "Ambush Sites"),
        "Close-quarters engagements" to RegionCoord("Close-quarters engagements", 38.9, -77.35, "Close-Quarters Engagements"),
        "Southern fortifications" to RegionCoord("Southern fortifications", 33.0, -87.0, "Southern Fortifications"),

        // ── Source / origin terms ─────────────────────────────────
        "Imported from Britain" to RegionCoord("Imported from Britain", 53.0, -1.5, "Imported from Britain"),
    )

    /**
     * Given a list of `whereFound` strings from a war relic, return
     * the matching coordinates for each one. Strings that don't match
     * the lookup are skipped — only confidently-placed regions are shown.
     */
    fun getRegionCoords(whereFound: List<String>): List<RegionCoord> {
        val results = mutableListOf<RegionCoord>()
        val seen = mutableSetOf<String>()
        for (region in whereFound) {
            val coord = REGION_COORDS[region] ?: continue
            if (seen.add(coord.label)) {
                results.add(coord)
            }
        }
        return results
    }
}
